index.d.ts

```diff
@@ -1618,18 +1618,68 @@ declare module 'mongoose' {
     validator: ValidateFn<T> | LegacyAsyncValidateFn<T> | AsyncValidateFn<T>;
   }
 
-  class VirtualType {
+  interface VirtualTypeOptions<D = Document> {
+    /** If `ref` is not nullish, this becomes a populated virtual. */
+    ref?: string | Function;
+
+    /**  The local field to populate on if this is a populated virtual. */
+    localField?: string | Function;
+
+    /** The foreign field to populate on if this is a populated virtual. */
+    foreignField?: string | Function;
+
+    /**
+     * By default, a populated virtual is an array. If you set `justOne`,
+     * the populated virtual will be a single doc or `null`.
+     */
+    justOne?: boolean;
+
+    /** If you set this to `true`, Mongoose will call any custom getters you defined on this virtual. */
+    getters?: boolean;
+
+    /**
+     * If you set this to `true`, `populate()` will set this virtual to the number of populated
+     * documents, as opposed to the documents themselves, using `Query#countDocuments()`.
+     */
+    count?: boolean;
+
+    /** Add an extra match condition to `populate()`. */
+    match?: FilterQuery<D> | ((doc: D) => FilterQuery<D>);
+
+    /** Add a default `limit` to the `populate()` query. */
+    limit?: number;
+
+    /** Add a default `skip` to the `populate()` query. */
+    skip?: number;
+
+    /**
+     * For legacy reasons, `limit` with `populate()` may give incorrect results because it only
+     * executes a single query for every document being populated. If you set `perDocumentLimit`,
+     * Mongoose will ensure correct `limit` per document by executing a separate query for each
+     * document to `populate()`. For example, `.find().populate({ path: 'test', perDocumentLimit: 2 })`
+     * will execute 2 additional queries if `.find()` returns 2 documents.
+     */
+    perDocumentLimit?: number;
+
+    /** Additional options like `limit` and `lean`. */
+    options?: QueryOptions;
+  }
+
+  class VirtualType<D = Document> {
+    /** VirtualType constructor */
+    constructor(options?: VirtualTypeOptions);
+
     /** Applies getters to `value`. */
-    applyGetters(value: any, doc: Document): any;
+    applyGetters(value: any, doc: D): any;
+
     /** Applies setters to `value`. */
-    applySetters(value: any, doc: Document): any;
+    applySetters(value: any, doc: D): any;
 
     /** Adds a custom getter to this virtual. */
-    // eslint-disable-next-line @typescript-eslint/ban-types
-    get(fn: Function): this;
+    get(fn: (this: D, value: any, virtual: this, doc: D) => any): this;
```

Comment:
Unfortunately I had to undo these changes. For some reason adding this generic param breaks defining `new Schema()` without a doc type and then using that schema for [`discriminator()`](https://mongoosejs.com/docs/api/model.html#model_Model.discriminator).